grpcbp/client_middlewares.go

```diff
@@ -0,0 +1,65 @@
+package grpcbp
+
+import (
+	"context"
+	"strings"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/reddit/baseplate.go/ecinterface"
+	"github.com/reddit/baseplate.go/tracing"
+	"google.golang.org/grpc"
+)
+
+type MonitorClientArgs struct {
+	ServiceSlug string
+}
+
+func MonitorInterceptor(args MonitorClientArgs) grpc.UnaryClientInterceptor {
+	prefix := args.ServiceSlug + "."
+	return func(
+		ctx context.Context,
+		method string,
+		req interface{},
+		reply interface{},
+		cc *grpc.ClientConn,
+		invoker grpc.UnaryInvoker,
+		opts ...grpc.CallOption,
+	) (err error) {
+		split := strings.Split(method, "/")
+		m := split[len(split)-1]
+		span, ctx := opentracing.StartSpanFromContext(
+			ctx,
+			prefix+m,
+			tracing.SpanTypeOption{
+				Type: tracing.SpanTypeClient,
+			},
+		)
+		ctx = CreateGRPCContextFromSpan(ctx, tracing.AsSpan(span))
+		defer func() {
+			span.FinishWithOptions(tracing.FinishOptions{
+				Ctx: ctx,
+				Err: err,
+			}.Convert())
+		}()
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+}
+
+func ForwardEdgeRequestContext(ecImpl ecinterface.Interface) grpc.UnaryClientInterceptor {
+	if ecImpl == nil {
+		ecImpl = ecinterface.Get()
+	}
```

Comment:
This has bit me before.  If nil is not a valid `ecinterface.Interface` (which it isn't) then it should be an error to pass it as one.  There should be exactly one place (presumably baseplate.New) that applies things like default values, if any.

Reply:
For me it's at least a questionable side effect. Should we just remove it?